swaybar/tray/menu.c

```diff
@@ -0,0 +1,975 @@
+	if (!(tray && tray->popup && tray->popup->pointer_focus)) {
+		return false;
+	}
+
+	struct swaybar_popup *popup = tray->popup;
+	double y = seat->pointer.y * popup->output->scale;
+	struct swaybar_popup_hotspot **hotspots_start =
+		(struct swaybar_popup_hotspot **)popup->pointer_focus->hotspots->items;
+	struct swaybar_popup_hotspot **hotspot_ptr = hotspots_start;
+	int step = 1;
+	if (popup->last_hover) { // calculate whether pointer went up or down
+		hotspot_ptr = popup->last_hover;
+		step = y < (*hotspot_ptr)->y ? -1 : 1;
+		hotspot_ptr += step;
+	}
+
+	for (; hotspot_ptr >= hotspots_start; hotspot_ptr += step) {
+		if ((step == 1) == (y < (*hotspot_ptr)->y)) {
+			break;
+		}
+	}
+
+	hotspot_ptr += step == -1;
+	struct swaybar_menu_item *item = (*hotspot_ptr)->item;
+	if (hotspot_ptr != popup->last_hover && item->enabled && !item->is_separator) {
+		struct swaybar_sni *sni = item->sni;
+		sd_bus_call_method_async(tray->bus, NULL, sni->service,
+				sni->menu_path, menu_interface, "Event", NULL, NULL, "isvu",
+				item->id, "hovered", "y", 0, time(NULL));
+		sway_log(SWAY_DEBUG, "%s%s hovered id %d", sni->service, sni->menu_path, item->id);
+	}
+	popup->last_hover = hotspot_ptr;
+
+	return true;
+}
+
+bool popup_pointer_button(void *data, struct wl_pointer *wl_pointer,
+		uint32_t serial, uint32_t time_, uint32_t button, uint32_t state) {
+	struct swaybar_seat *seat = data;
+	struct swaybar_tray *tray = seat->bar->tray;
+	if (!(tray && tray->popup)) {
+		return false;
+	}
+
+	struct swaybar_popup *popup = tray->popup;
+	float y = seat->pointer.y * popup->output->scale;
+	if (state != WL_POINTER_BUTTON_STATE_PRESSED) {
+		// intentionally left blank
+	} else if (!popup->pointer_focus) {
+		close_popup(popup);
+	} else if (button == BTN_LEFT) {
+		list_t *hotspots = popup->pointer_focus->hotspots;
+		for (int i = 0; i < hotspots->length; ++i) {
+			struct swaybar_popup_hotspot *hotspot = hotspots->items[i];
+			if (y < hotspot->y) {
+				struct swaybar_menu_item *item = hotspot->item;
+
+				if (!item->enabled || item->is_separator) {
+					break;
+				}
+
+				struct swaybar_sni *sni = popup->sni;
+				if (item->children) {
+					destroy_popup_surface(popup->pointer_focus->child);
+					popup->pointer_focus->child = NULL;
+					popup->serial = serial;
+					popup->x = 0;
+					if (tray->bar->config->position & ZWLR_LAYER_SURFACE_V1_ANCHOR_TOP) { // top bar
+						popup->y = i ? ((struct swaybar_popup_hotspot *)hotspots->items[i-1])->y : 0;
+					} else {
+						popup->y = hotspot->y;
+					}
+					popup->y /= popup->output->scale;
+					open_popup_id(sni, item->id);
+				} else {
+					sd_bus_call_method_async(tray->bus, NULL, sni->service,
+							sni->menu_path, menu_interface, "Event", NULL, NULL,
+							"isvu", item->id, "clicked", "y", 0, time(NULL));
+					sway_log(SWAY_DEBUG, "%s%s popup clicked id %d",
+							sni->service, sni->menu_path, item->id);
+					close_popup(popup);
+				}
+				break;
+			}
+		}
```

Comment:
This seems like a good idea, I'll have a look into it.

Which applications have different menus?

Reply:
`nm-applet` does something different under `i3`/X when not using `--indicator` option; it presents different left vs right click menus.  With `--indicator` (`i3`/X or `sway`/otherwise), the menus are the same.

Reply:
That's also the case for Transmission, without libappindicator, left-clicking the icon toggles the main window.
As far as I know this is a libappindicator limitation, and the protocol actually does distinguish between left and right click.

Reply:
The protocol itself has some limitations if used with wayland. All clicks that do not show the dbusmenu send coordinates so an (Xorg) application can open a window at that location.

As a result the left click (Activate) can only be reasonably used for a simple task like showing/hiding the main window or toggling mute in wayland
There are also options for middle click (SecondaryActivate) and scroll events (Scroll), for example for volume control, headphone switch, ...
A right click shows the dbusmenu if it is available, otherwise the ContextMenu method is called. 
The ContextMenu method is also not a good idea for wayland since that relies on positioning a window at specific coordinates as well.
When ItemIsMenu is set, left click acts exactly the same as right click.

libappindicator doesn't implement any support for Activate and ContextMenu events. SecondaryActivate can only activate a visible and sensitive GTK menuitem. Scroll events seem to be handled properly.
In my opinion only dbusmenu for all clicks and scroll events can used with libappindicator. For anything else that library has to be improved or replaced.